### DISCUSSION TITLE
Add binary release package for yarp github release, and add necessary powershell script to the package.

### DISCUSSION THREAD

**Discussion by kasini3000** [dotnet/yarp#1404] [Ideas] [answered]
### Current
I said that yarp did not work out of the box. It's like, you buy a car, but you have to crank it with a lever. This is very unprofessional. This greatly limits the widespread use of yarp.

According to the tutorial, you must:
> install .net6 runtime, then Update Program.cs to use the YARP to start it:

```c#
var builder = WebApplication.CreateBuilder(args);
builder.Services.AddReverseProxy()
     .LoadFromConfig(builder.Configuration.GetSection("ReverseProxy"));
var app = builder.Build();
app.MapReverseProxy();
app.Run();
```

# Subject: Add binary release package for yarp github release, and add necessary powershell script to the package.

### Subtitle 1: powershell is the best solution for win and linux deployment of yarp.

* Let yarp depend powershell v7.2 . powershell 7.2 is based on .net6, which can provide the maximum performance of yarp.
Only .net 6 binary on the binary package of release. (No change on nuget, .net 3.1; .net5; .net6 provided)

* powershell is cross-platform. Can provide command line parameters. Control scripts can be easily developed in the future.

* Powershell supports json, yaml (via modules).

* There is already most linux one-click to install powershell project (shell script) .
https://gitee.com/chuanjiao10/kasini3000_agent_linux

### Subtitle 2: The ps1 script should not be included in the nuget package, but a powershell script should be added to the binary release package.

* Does not need to be released some-control.exe

* The nuget package is only suitable for .net program applications and not suitable for direct deployment in linux production.

###  Subtitle 3: Control yarp

* You can control yarp by calling yarp's API methods and property through powershell scripts.

* You can use powershell to call yaml, json, which is the method of configuration files, to control yarp.

* Deploy to win and linux: no need to install .net6 runtime, no need to write c# files.

**Comment by Tratcher** [comment, marked as answer]
Related: https://github.com/microsoft/reverse-proxy/issues/167, https://github.com/microsoft/reverse-proxy/issues/261

The library approach is intentional, it's expected that many consumers of YARP want to customize it with code. While we may provide pre-built images in the future, there's currently little reason to use that compared to other reverse proxies.

Can you create an example of what it would be like to use YARP from Powershell? That's not a common approach we've seen with ASP.NET Core based libraries.

**Comment by kasini3000** [comment]
My environment is win+powershell7.2 (.net6),
My start yarp code on these yarp manuals does not work. It needs minor modifications. need some help.
```ps
Set-Location $PSScriptRoot # into current directory
Add-Type -Path "${PSScriptRoot}/net6.0/Yarp.ReverseProxy.dll" # download from nuget
Add-Type -Path "${PSScriptRoot}/ReverseProxy.cs"
#or
Add-Type @'
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddReverseProxy()
    .LoadFromConfig(builder.Configuration.GetSection("ReverseProxy"));

var app = builder.Build();

app.MapReverseProxy();

app.Run();
'@
```

**Reply by Tratcher** [reply]
I tried this and hit multiple errors that I'm not sure how to get past.

```
(1,1): error CS8805: Program using top-level statements must be an executable. var builder =
     | WebApplication.CreateBuilder(args); ^
(1,15): error CS0103: The name 'WebApplication' does not exist in the current context var builder =
     | WebApplication.CreateBuilder(args);               ^
(1,17): error CS0234: The type or namespace name 'AspNetCore' does not exist in the namespace
     | 'Microsoft' (are you missing an assembly reference?) using Microsoft.AspNetCore.Builder;
```
For the first one I added the Program.Main around it, but not sure how well that will work.
```c#
using Microsoft.AspNetCore.Builder;

public static class Program
{
  public static void Main(string[] args)
  {
    var builder = WebApplication.CreateBuilder(args);

    builder.Services.AddReverseProxy()
      .LoadFromConfig(builder.Configuration.GetSection("ReverseProxy"));

    var app = builder.Build();

    app.MapReverseProxy();

    app.Run();
  }
}
```
I think I need the namespace and `<FrameworkReference Include="Microsoft.AspNetCore.App" />` to address the second and third issues, but I'm not sure where/how to add a FrameworkReference in powershell.

**Reply by kasini3000** [reply]
First of all, I think the important questions are:
1 Whether yarp wants to be controlled by powershell.
2 How to control? Whether to call powershell within the yarp process, or let yarp be a child process of powershell. And how they communicate with each other. (Pass .net objects, properties, methods)
After the above problem is determined,we need to find the powershell team or fans to cooperate.

----
As for  **and FrameworkReference Include="Microsoft.AspNetCore.App"**
  I think we should do this:
Download the necessary dll in Microsoft.AspNetCore.App from nuget, put it into yarp release package, and use a method similar to Add-Type -Path "${PSScriptRoot}/net6.0/Microsoft.AspNetCore.dll" for reference.

Find and download the dll:
https://www.nuget.org/packages/Microsoft.AspNetCore.App/

**Reply by Tratcher** [reply]
Note https://www.nuget.org/packages/Microsoft.AspNetCore.App/ is obsolete since 3.0. The next best option is downloading the zips from https://github.com/dotnet/installer#installers-and-binaries.

**Reply by Tratcher** [reply]
After talking with the team I don't think running asp.net code from within powershell is possible/supported, the libraries just aren't available. That leaves using powershell to control an existing YARP based app. You could modify the config file, or call into a control API depending on what you wanted to accomplish. This might tie into the design for #261 (pre-built-exe).

**Reply by kasini3000** [reply]
OK, I see. Thank you for your feedback.

**Comment by kasini3000** [comment]
![car2](https://user-images.githubusercontent.com/65390418/142983171-6f514d88-f0d5-46e8-916c-5afdfef23151.jpeg)
